### PR TITLE
Persist NTP-SI index and for both Rewards and non-Rewards users

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/30010

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

- Confirm on fresh install the first NTT ad is shown on the 2nd new tab
- Confirm subsequent NTT ads are shown every 4th new tab
- Confirm browser relaunches persist the the count between new tabs
- Confirm restored tabs (non visible) do not increment the count after relaunching the browser
- Confirm opening a new tab and relaunching the browser many times shows an NTT ad every nth tab

### CODE REVIEW

File: browser/ui/webui/new_tab_page/brave_new_tab_ui.cc [hunk 1]
```diff
@@ -35,6 +37,13 @@ BraveNewTabUI::BraveNewTabUI(content::WebUI* web_ui, const std::string& name)
           web_ui,
           true /* Needed for legacy non-mojom message handler */),
       page_factory_receiver_(this) {
+  const bool was_restored =
+      web_ui->GetWebContents()
+          ->GetController()
+          .GetLastCommittedEntry()
+          ->IsRestored() &&
+      web_ui->GetWebContents()->GetVisibility() != content::Visibility::VISIBLE;
```
Comment (tmancey) [inline browser/ui/webui/new_tab_page/brave_new_tab_ui.cc:45]: @iambrianfung when restoring new tab pages on browser launch, if the active tab is a blank new tab page should we rotate the index? If not I will remove this change, and if so I will refactor this change

File: components/ntp_background_images/browser/view_counter_service.cc [hunk 2]
```diff
@@ -57,6 +57,7 @@ void ViewCounterService::RegisterProfilePrefs(
     user_prefs::PrefRegistrySyncable* registry) {
   registry->RegisterBooleanPref(
       prefs::kBrandedWallpaperNotificationDismissed, false);
+  registry->RegisterIntegerPref(prefs::kCountToBrandedWallpaper, 1);
```
Comment (tmancey) [inline components/ntp_background_images/browser/view_counter_service.cc:60]: We could fetch the initial value from the model, @simonhong your thoughts?

Comment (simonhong) [reply components/ntp_background_images/browser/view_counter_service.cc:60]: Yup, model could give initial value.

**Review by simonhong — COMMENTED** [review]
~~We have to reset cached count when component is updated as each campaign could have different number of images.~~

NVM - this count is not related with each campaign's image numbers.

**Comment by simonhong** [comment]
Tested this PR and it doesn't work as expected.
Re-launched but newtab doesn't show brand. Before relaunching, newtab has brand image.

When NTP shows brand image, count is already set to `3` because we update it for the next ntp loading.
If it's `0`, it means next NTP loading will show brand. IIRC, service provide brand image to NTP baesd on that count,
and then that count is decreased. 
Screenshot 2023-05-10 at 11 03 35 AM

**Comment by tmancey** [comment]
@simonhong before this change PR, branded image is not remembered when relaunching the browser "Tested this PR and it doesn't work as expected.

File: components/ntp_background_images/browser/view_counter_service.cc [hunk 3]
```diff
@@ -91,6 +95,9 @@ ViewCounterService::ViewCounterService(
 
   ResetModel();
 
+  model_.set_count_to_branded_wallpaper(
+      prefs_->GetInteger(prefs::kCountToBrandedWallpaper));
```
Comment (simonhong) [inline components/ntp_background_images/browser/view_counter_service.cc:99]: How about doing this in the model's ctor?

File: browser/ui/webui/new_tab_page/brave_new_tab_ui.cc [hunk 4]
```diff
@@ -35,6 +36,14 @@ BraveNewTabUI::BraveNewTabUI(content::WebUI* web_ui, const std::string& name)
           web_ui,
           true /* Needed for legacy non-mojom message handler */),
       page_factory_receiver_(this) {
+  bool was_restored = web_ui->GetWebContents()
```
Comment (aseren) [inline browser/ui/webui/new_tab_page/brave_new_tab_ui.cc:39]: nit: 
```
WebContents* web_contents = web_ui->GetWebContents();
CHECK(web_contents);
bool was_restored = web_contents->GetController()
                          .GetLastCommittedEntry()
                          ->IsRestored();
```

File: browser/ui/webui/new_tab_page/brave_new_tab_ui.cc [hunk 5]
```diff
@@ -35,6 +38,15 @@ BraveNewTabUI::BraveNewTabUI(content::WebUI* web_ui, const std::string& name)
           web_ui,
           true /* Needed for legacy non-mojom message handler */),
       page_factory_receiver_(this) {
+  content::WebContents* web_contents = web_ui->GetWebContents();
+  CHECK(web_contents);
+
+  const bool was_restored =
+      web_contents->GetController().GetLastCommittedEntry()->IsRestored();
```
Comment (aseren) [inline browser/ui/webui/new_tab_page/brave_new_tab_ui.cc:45]: I'm not sure about GetLastCommittedEntry() value in the BraveNewTabUI constructor. Maybe it is safer to check its value by
```
if (const auto* last_committed_entry = web_contents->GetController().GetLastCommittedEntry()) {
  was_restored = last_committed_entry->IsRestored();
}

```

File: components/ntp_background_images/browser/view_counter_model.h [hunk 6]
```diff
@@ -65,7 +71,8 @@ class ViewCounterModel {
   void RegisterPageViewForBackgroundImages();
 
   // For NTP SI.
-  int count_to_branded_wallpaper_ = 0;
+  raw_ptr<PrefService> prefs_ = nullptr;
```
Comment (aseren) [inline components/ntp_background_images/browser/view_counter_model.h:74]: raw_ref<PrefService>

**Review by simonhong — APPROVED** [review]
++ 👍🏼 
and please set STR about restoring fix in the issue for QA.

**Comment by aseren** [comment]
LGTM with some comments

**Review by aseren — APPROVED** [review]

**Pull request merged**